### PR TITLE
Optional instance support

### PR DESCRIPTION
**Pull request by SCdF**
Actions can now declare they don't need an API instance to function.

Currently only implemented for initialise-project-layout, but could very
easily be done on the others where it matters.

Issue: medic/medic-conf#218

### CODE REVIEW

File: src/cli/supported-actions.js [hunk 1]
```diff
@@ -1,27 +1,7 @@
...
-  'upload-collect-forms',
-  'upload-contact-forms',
-  'upload-custom-translations',
-  'upload-docs',
-  'upload-resources',
-  'upload-sms-from-csv',
-];
+const fs = require('fs');
+
+const actions = fs.readdirSync(`${__dirname}/../fn`)
```
Comment (SCdF) [inline src/cli/supported-actions.js]: Automating this removes another place you need to check. The only change was there was also a secret "compress-images" action that does both image types..

**Comment by SCdF** [comment]
NB: this only does the action I was interested in. If we're OK with this code I can raise another issue to go through and work out which of the other actions don't need an instance and change their APIs as well

**Comment by kennsippell** [comment]
This is tracked by https://github.com/medic/medic-conf/issues/218. Can you make sure you close that issue when this merges?

File: test/lib/main.spec.js [hunk 2]
```diff
@@ -87,6 +87,12 @@ describe('main', () => {
     expect(main.__get__('process').env.NODE_TLS_REJECT_UNAUTHORIZED).to.eq('0');
   });
 
+  it('supports actions that do not require an instance', async () => {
```
Comment (kennsippell) [inline test/lib/main.spec.js]: Do we test the case that the instance data is still required for other actions? The "'Failed to obtain a url to the API'" case?

File: src/lib/main.js [hunk 3]
```diff
@@ -175,4 +178,23 @@ module.exports = async (argv, env) => {
   }
 };
 
-const executeAction = action => require(`../fn/${action}`)();
+
+const requiresInstance = actions => actions.find(actionName => {
+  const actionLib = require(`../fn/${actionName}`);
+  if (typeof actionLib === 'function') {
+    return true;
+  } else if (typeof actionLib.requiresInstance === 'boolean') {
```
Comment (kennsippell) [inline src/lib/main.js]: the `else` is not required after `if() { return }` blocks

Comment (kennsippell) [inline src/lib/main.js]: if it isn't a function, should we just assume it is an object and accept any truthy value for `requiresInstance`?

File: src/lib/main.js [hunk 4]
```diff
@@ -175,4 +178,23 @@ module.exports = async (argv, env) => {
   }
 };
 
-const executeAction = action => require(`../fn/${action}`)();
+
+const requiresInstance = actions => actions.find(actionName => {
+  const actionLib = require(`../fn/${actionName}`);
```
Comment (kennsippell) [inline src/lib/main.js]: We also `require()` each action file in the `executeAction` function. Can we refactor so the action files are only required once?

File: src/fn/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Actions
+
+Actions are defined as usable if they exist in this directory.
+
+The API is that they should export something like this:
+
+```js
+module.exports = {
+    requiresInstance: true,
+    execute: async (projectDir, couchUrl, extraArgs) => {
```
Comment (kennsippell) [inline src/fn/README.md]: It seems you call this `execute()` here in the documents but `action()` in the code. I prefer your choice of execute. A little confusing to use `action` to describe both the files and a function within them.

File: src/fn/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+**OR** just:
+
+```js
+module.exports = async (projectDir, couchUrl, extraArgs) => {
+    ...
+}
+```
+
+In which case defaults will be used.
```
Comment (kennsippell) [inline src/fn/README.md]: nit: a table might work well here. with columns parameter, description, required, default

**Comment by garethbowen** [comment]
Looks like Kenn is on top of this one. I gave it a quick look and the concept makes sense to me. At some point I'd refactor all actions to export an object to standardise the API.

File: test/lib/main.spec.js (hunk 2 above)
Comment (SCdF) [reply test/lib/main.spec.js]: I presumed we did already (since I didn't add that code flow), though I didn't look that hard. I'll take another look and add it if it's not there.

File: src/lib/main.js (hunk 3 above)
Comment (SCdF) [reply src/lib/main.js]: Yeah I know, but IMO when all branches are "logically equal" (like one isn't an edge case or whatever, they're all valid) it looks clearer to have the return statements lined up vertically, as it puts those forks mentally on an equal footing. I realise this is A E S T H E T I C S territory, let me know if you're strongly against the redundancy.

Comment (SCdF) [reply src/lib/main.js]: No, because if you don't define it that will default to falsey, which is not what we want IMO.

We could either change the wording to be the opposite (`noInstanceNeeded`) to drop this logic block, or we could go through all of the existing actions and correct them into this format, allowing us to change the API to requiring you to declare that you need an instance before you're guaranteed one. However, I think defaulting to presuming you need an instance makes more sense unless we want to refactor _even further_ to not allowing actions to access instances without first declaring that they need it.

File: src/fn/README.md (hunk 5 above)
Comment (SCdF) [reply src/fn/README.md]: Whoops, yeah it's supposed to be `execute`, I'll change it

File: src/fn/README.md (hunk 6 above)
Comment (SCdF) [reply src/fn/README.md]: okey dokey

File: src/lib/main.js (hunk 4 above)
Comment (SCdF) [reply src/lib/main.js]: Well so the good news is that Node does cache these calls so they aren't actually doing anything too inefficient, but I agree it's a bit pants.

Can we address this in a later refactor? I didn't want to get too sidetracked, but alongside this there is quite a bit of cleanup to do around actions (their should be a module that deals with them instead of the code being dumped at the bottom of main, we shouldn't pass as couchUrl to actions that don't need it, etc)

File: src/lib/main.js (hunk 3 above)
Comment (kennsippell) [reply src/lib/main.js]: This is actually one of the things we call out explicitly in our style guide. https://github.com/medic/medic-docs/blob/master/development/style-guide.md#return-early-from-functions

File: test/lib/main.spec.js (hunk 2 above)
Comment (kennsippell) [reply test/lib/main.spec.js]: i think i checked before leaving the comment and it wasn't present

File: src/lib/main.js (hunk 3 above)
Comment (kennsippell) [reply src/lib/main.js]: I suggested "any defined truthy value" so `return actionLib.requiresInstance !== undefined && !!actionLib.requiresInstance;`

File: src/lib/main.js (hunk 4 above)
Comment (kennsippell) [reply src/lib/main.js]: > we shouldn't pass as couchUrl to actions that don't need it, etc

Note that we don't pass a couchUrl to actions anymore

File: src/fn/README.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    ...
+}
+```
+
+## Module Exports
+
+|Field|Required|Notes|
+|---|---|---|
+|`requiresInstance`|Optional, defaults to `true`|The action needs the user to have provided a instance location, e.g. via `--local` or `--instance`|
+|`execute`|Required|the function that is run when the action is executed. Is passed `projectDir`, `couchUrl` and any `extraArgs`|
```
Comment (kennsippell) [inline src/fn/README.md]: > Is passed projectDir, couchUrl and any extraArgs

This is no longer true

File: package.json [hunk 8]
```diff
@@ -10,7 +10,7 @@
     "clean": "rm -rf ./build/",
     "generate-changelog": "echo 'Generating changelog...' && ./scripts/changelog > CHANGELOG.md && echo 'Changelog generated.'",
     "eslint": "eslint src/**/*.js test/*.js test/**/*.js",
-    "test": "npm run eslint && npm run clean && mkdir -p build/test && cp -r test/data build/test/data && cd build/test && mocha ../../test/**/*.spec.js",
+    "test": "npm run clean && mkdir -p build/test && cp -r test/data build/test/data && cd build/test && mocha ../../test/**/*.spec.js",
```
Comment (kennsippell) [inline package.json]: why this change? i believe npm test is used by travis

File: src/lib/main.js [hunk 9]
```diff
@@ -175,4 +178,23 @@ module.exports = async (argv, env) => {
...
+  } else {
+    return true;
+  }
+});
+
+const executeAction = actionName => {
+  const actionLib = require(`../fn/${actionName}`);
+  if (typeof actionLib === 'function') {
+    return actionLib();
+  } else {
```
Comment (kennsippell) [inline src/lib/main.js]: one additional example of `if () { return } else {`

File: src/lib/main.js (hunk 3 above)
Comment (SCdF) [reply src/lib/main.js]: :+1:

File: package.json (hunk 8 above)
Comment (SCdF) [reply package.json]: Whoops I didn't mean to merge that, I'll revert

File: src/lib/main.js (hunk 3 above)
Comment (SCdF) [reply src/lib/main.js]: I think that would mean we'd be supporting truthy but not falsy (because having `requiresInstance` as `undefined` wouldn't distinguish between that being an intended falsy value and it not being present at all) which seems confusing. 

I've changed it to what I should have done in the first place, where it probes the object and if the key is present it uses it.

File: src/lib/main.js [hunk 10]
```diff
@@ -175,4 +178,25 @@ module.exports = async (argv, env) => {
...
 
-const executeAction = action => require(`../fn/${action}`)();
+
+const requiresInstance = actions => actions.find(actionName => {
+  const actionLib = require(`../fn/${actionName}`);
+  if (typeof actionLib === 'function') {
+    return true;
+  }
+
+  if (Object.keys(actionLib).includes('requiresInstance')) {
```
Comment (kennsippell) [inline src/lib/main.js]: nit: or `actionLib.hasOwnProperty('requiresInstance')`

File: src/lib/main.js (hunk 3 above)
Comment (kennsippell) [reply src/lib/main.js]: Nice! `actionLib.hasOwnProperty('requiresInstance')` is also a good option here.

Comment (SCdF) [reply src/lib/main.js]: Aye, I forgot about that function, I think that's a clearer function usage, I'll swap it in :+1:

File: src/lib/main.js [hunk 11]
```diff
@@ -160,43 +175,23 @@ module.exports = async (argv, env) => {
   // GO GO GO
   //
   info(`Processing config in ${projectName}.`);
-  info('Actions:\n     -', actions.join('\n     - '));
+  info('Actions:\n     -', actions.map(({name}) => name).join('\n     - '));
```
[the remark this answers is not in the document]
Comment (SCdF) [reply src/lib/main.js]: https://xkcd.com/224/

**Comment by SCdF** [comment]
Merging, as AT is happening as part of the tutorial I'm writing which instigated the change
